src/v/security/credential_store.h

```diff
@@ -63,9 +66,25 @@ class credential_store {
         return _credentials.contains(name);
     }
 
-    const_iterator begin() const { return _credentials.cbegin(); }
-    const_iterator end() const { return _credentials.cend(); }
+    // Ephemeral credentials often require careful handling; they must not
+    // be serialized to disk, and in the general case, should not be displayed
+    // to users.
+    static constexpr auto is_not_ephemeral =
+      [](security::credential_store::container_type::value_type const& t) {
+          return ss::visit(t.second, [](security::scram_credential const& c) {
+              return !c.principal().has_value()
+                     || c.principal().value().type()
+                          != security::principal_type::ephemeral_user;
+          });
+      };
 
+    // Retrieve a list of credentials that satisfy the predicate.
+    //
+    // E.g.:
+    // _creds.range(credential_store::is_not_ephemeral);
+    auto range(auto pred) {
+        return boost::adaptors::filter(_credentials, std::move(pred));
+    }
```

Comment:
Can we include a default `range()` that automatically filters out ephemeral users? Otherwise devs must remember to pass in the `is_not_ephemeral` predicate.

Reply:
The whole point is to make the thought process explicit.

Reply:
I understand what you're saying. However, it feels like it'll be easy for a dev to put ephemeral creds in a place they shouldn't which is what happened in #11141 

If this code change were to merge, then what stops someone from doing:
```cpp
_creds.range();
```

Reply:
I saw it as I wrote it out, the above would be a compile error. I'll come back to this later to re-review.

Reply:
There are a lot of potential interface opportunities, such as a `for_each-credential`, for example, which I dislike,

There's at least one example of using `reserve` with `std::distance`, which may be  sub-optimal with a forward iterator.